OpenRA.Game/Primitives/RotatingBuffer.cs

```diff
@@ -0,0 +1,141 @@
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+
+namespace OpenRA.Primitives
+{
+	/// <summary>Fixed size rorating buffer backed by an array.</summary>
+	public class RotatingBuffer<T> : ICollection<T>, IEnumerable<T>
+	{
+		readonly IComparer<T> comparer;
+		readonly T[] values;
+		int start;
+		int count;
+
+		public int Capacity => values.Length;
+		public int Count => count;
+		public bool IsReadOnly => false;
+
+		public RotatingBuffer(int capacity, IComparer<T> comparer)
+		{
+			this.comparer = comparer;
+			values = new T[capacity];
+			start = 0;
+			count = 0;
+		}
+
+		public RotatingBuffer(int capacity)
+			: this(capacity, Comparer<T>.Default) { }
+
+		public void Add(T value)
+		{
+			values[(start + count) % values.Length] = value;
+			if (count < values.Length)
+				count++;
+			else
+				start = (start + 1) % values.Length;
+		}
+
+		public void Clear()
+		{
+			Array.Clear(values, 0, values.Length);
+			count = 0;
+		}
+
+		public bool Contains(T value)
+		{
+			var capacity = values.Length;
+			var end = start + count;
+			for (var i = start; i < end; ++i)
+				if (comparer.Compare(values[i % capacity], value) == 0)
+					return true;
+
+			return false;
+		}
+
+		public void CopyTo(T[] array, int arrayIndex)
+		{
+			if (array == null)
+				throw new ArgumentNullException("Target array is null");
+
+			if (arrayIndex < 0)
+				throw new ArgumentNullException("Index is less than 0");
+
+			if (arrayIndex + count >= array.Length)
+				throw new ArgumentException("Invalid array capactiy");
+
+			var di = arrayIndex;
+			var end = start + count;
+			var capacity = values.Length;
+			for (var i = start; i < end; ++i)
+				array[di++] = values[i % capacity];
+		}
+
+		public bool Remove(T value)
+		{
+			var capacity = values.Length;
+			var end = start + count;
+			for (var i = start; i < end; ++i)
+				if (comparer.Compare(values[i % capacity], value) == 0)
+				{
+					end--;
+					for (var j = i; j < end; ++j)
+						values[j % capacity] = values[(j + 1) % capacity];
+
+					count--;
+					return true;
+				}
+
+			return false;
+		}
+
+		public T this[int pos]
+		{
+			get => values[(start + pos) % values.Length];
+
+			set
+			{
+				if (pos >= count)
+					throw new ArgumentException($"Index out of bounds: {pos}");
+
+				values[(start + pos) % values.Length] = value;
+			}
+		}
+
+		public T First()
+		{
+			if (count == 0)
+				throw new ArgumentException("Empty buffer");
+
+			return values[start];
+		}
+
+		public T Last()
+		{
+			if (count == 0)
+				throw new ArgumentException("Empty buffer");
+
+			return values[(start + count - 1) % values.Length];
+		}
+
+		public IEnumerator<T> GetEnumerator()
+		{
+			var c = count;
+			var index = start;
```

Comment:
Shouldn't this be 0?

Reply:
No, it should be the start/head of the buffer.

Reply:
This won't work with the `while (index < c)` and `(start + index++)` below.

Reply:
Why? I does a mod (%) size to make it wrap?

It loops the size, independant of the start position. Then after select the element from the start position - but making sure to mod it by the size?